org.eclipse.cdt.cross.arc.gnu/src/org/eclipse/cdt/cross/arc/gnu/common/StateChecker.java

```diff
@@ -0,0 +1,163 @@
+/*******************************************************************************
+ * This program and the accompanying materials are made available under the terms of the Common
+ * Public License v1.0 which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/cpl-v10.html
+ *
+ * Copyright (c) 2016 Synopsys, Inc.
+ *******************************************************************************/
+
+package org.eclipse.cdt.cross.arc.gnu.common;
+
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.eclipse.cdt.cross.arc.gnu.common.ResourceChangeListener.ProjectEventListener;
+import org.eclipse.cdt.internal.core.WeakHashSet;
+import org.eclipse.cdt.make.core.scannerconfig.IScannerInfoCollector;
+import org.eclipse.cdt.make.core.scannerconfig.InfoContext;
+import org.eclipse.cdt.make.internal.core.scannerconfig2.PerProjectSICollector;
+import org.eclipse.core.resources.IMarker;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.ProjectScope;
+import org.eclipse.core.resources.WorkspaceJob;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.preferences.IEclipsePreferences;
+import org.eclipse.ui.statushandlers.StatusManager;
+
+/**
+ * This class is intended for checking the compiler plug-in state number (it may be different for
+ * the same plug-in version) and notifying user if his compiler plug-in has incompatible changes in
+ * comparison with a compiler plug-in a project was created with. It makes user aware of possible
+ * reason of his problems if they appear.
+ *
+ * When creating a new project it writes a compiler plug-in state number to the
+ * $PROJECT/.settings/com.arc.cdt.prefs, warns user when importing, building, debugging (once per
+ * project during a one IDE launch) a project created with an incompatible compiler plug-in state
+ * number.
+ */
+public class StateChecker {
+
+  // It is used to avoid duplicating warnings for the same project during the one IDE launch.
```

Comment:
Please rephrase. When statement starts with "It" I automatically start thinking that it speaks about the `@SuppressWarnings` stuff. In general you can refer to "It" only after you already mentioned something, but here comment goes before the code, so it should be something like "This field is used to avoid...". Even better would to explain mechanism of how it is used - "This field is used to store projects for which version check already has been done - we don't want to print warning message twice for same project". You can phrase differently if you want.